### DISCUSSION TITLE
Is it possible to apply undo-tree into git history using magit?

### DISCUSSION THREAD

**Discussion by avatar-lavventura** [magit/magit#4674] [Ideas]
As we know [undo-tree][1]: a package allowing unlimited undo and redo without ever breaking the chain thanks to its tree (rather than linear) structure

`undo-tree` is a great package that helps me to move around in the history of the open file. I was wondering could similiar approach applied into git history using magit? Basically I just want to move in the history of the currently opened file fastest way possible (could be with a single keybinding without a tree show up).

Please note that this may already handled my `magit` but I was not able to find a way.

I can apply `magit-status` , than select the corresponding commit than on the `commit ` window when I enter again any file name its version from the history shows up. But this flow I need the apply three steps.

I just want this process to be happen right away like in the undo-tree where I can go back in the git-history of the file.

  [1]: http://www.dr-qubit.org/undo-tree.html

**Comment by tarsius** [comment]
Press `C-c M-g` to bring up `magit-file-dispatch`. Then use `p` and `n`.

But it's not like undo-tree's history browsing and I don't intend to make any improvements any time soon.

**Comment by avatar-lavventura** [comment]
Can `magit-file-dispatch` and `p` called in a single keybinding?

**Reply by tarsius** [reply]
Of course.

**Reply by avatar-lavventura** [reply]
```lisp
(defun previous_history ()
  (interactive)
  (magit-file-dispatch)
  (p))
```

is this the way? I am not sure how I can use `p` as input

**Reply by kyleam** [reply]
`p` is bound to `magit-blob-previous` in the `magit-file-dispatch` transient. 

https://github.com/magit/magit/blob/3cb7f5ba430906bded9e5d9951f5260ab25644d0/lisp/magit-files.el#L308-L309

 You can bind the `magit-blob-previous` command to whatever binding you want.  And if you're in the working tree version of a file and call `magit-blob-previous`, the `magit-blob` minor mode will be active after the first call, and then `magit-blob-previous` is available on `p`.

https://github.com/magit/magit/blob/ab801de53827a232b7806362fb08ca804f27c6d0/lisp/magit-files.el#L326-L327
